### PR TITLE
Logger merged

### PR DESCRIPTION
**Pull request by MTBC-Go-MF56**

### CODE REVIEW

File: common/enums/decisions.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from enum import Enum
+
+
+class Decisions(Enum):
+    Accelerate = 1
+    De_accelerate = 2
```
Comment (Nabeel07925) [inline common/enums/decisions.py:6]: keep these in CamelCase

File: common/enums/model_types.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+from enum import Enum
+
+
+class ModelTypes(Enum):
+    Rule_based = 1
```
Comment (Nabeel07925) [inline common/enums/model_types.py:5]: Same

File: common/utility/driving/driving_calculations.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
+from common.enums.road_types import *
+from common.utility.driving.angle_calculator import AngleCalculator
+import numpy as np
+
+
+def point_to_line_intersection(point, neighbours):
```
Comment (internaccounts123) [inline common/utility/driving/driving_calculations.py:6]: Make a class and make all these functions static methods.

File: map/map_class.py [hunk 4]
```diff
@@ -54,17 +53,50 @@ def get_lateral_lanes(self, lane_id, road_name):
...
+            if road.road_id == road_id:
+                for lane_idx in range(len(road.lanes)):
+                    if road.lanes[lane_idx].id == l_id:
+                        if dec == "Move_right":
+                            lane_id = road.lanes[lane_idx+1].id
+                        elif dec == "Move_left":
+                            lane_id = road.lanes[lane_idx - 1].id
+                        else:
+                            lane_id = l_id
+
```
Comment (internaccounts123) [inline map/map_class.py:70]: return road id as well.

File: map/map_class.py [hunk 5]
```diff
@@ -54,17 +53,50 @@ def get_lateral_lanes(self, lane_id, road_name):
 
         return lateral_lanes
 
+    def update_lane_info(self, road_id, l_id, dec):
+        lane_id = None
+
+        for road in self.__roads:
```
Comment (internaccounts123) [inline map/map_class.py:60]: Road info should be independent from decision

File: map/map_class.py [hunk 6]
```diff
@@ -84,5 +116,22 @@ def check_point_fit(self, current_position, lane_points):
 
         return current_position[1] == slope * current_position[0] - intercept
 
-
-
+    def straight_road_info(self, road_id, lane_id):
+        arr = []
+        for i in range(len(self.__roads)):
```
Comment (internaccounts123) [inline map/map_class.py:122]: remove this function.

File: simulation/traffic/vehicle/rule_based/rule_based.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        if decision == "Lane_change":
+            return False
+        else:
+            return True
+
+    def make_decision(self, grid, lane_points, d_points, right_lane_points, right_d_points, right_car_list,
+                      left_lane_points, left_d_points, left_car_list):
+        current_road = grid[self.road_id][self.lane_id]
+        two_sec_decision = two_sec_rule(self, current_road, lane_points, d_points)
+        margin_point = self.speed_limit - (self.speed_limit * .01)
```
Comment (internaccounts123) [inline simulation/traffic/vehicle/rule_based/rule_based.py:61]: initialize this percentage in init

File: simulation/traffic/vehicle/vehicle_class.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    @current_acc.setter
+    def current_acc(self, current_acc):
+        self.__current_acc = current_acc
+
+    def move(self, decision, lane_points, right_lane_points, left_lane_points):
+        _neigh_1, _neigh_2 = get_neighbouring_points(lane_points, [self.x, self.y])
+        bearing = AngleCalculator.get_bearing(_neigh_1[0], _neigh_2[0])
+
+        if Decisions[decision].value == Decisions.Accelerate.value:
+            self.speed += (self.current_acc * (1.0/ConfigReader.get_data("fps")[0]))
```
Comment (internaccounts123) [inline simulation/traffic/vehicle/vehicle_class.py:168]: move position calculation to drive calculations

File: simulation/world/world.py [hunk 9]
```diff
@@ -37,46 +48,44 @@ def world_map (self, world_map ):
         self.__world_map = world_map
 
     def update(self, event):
-        for i in range(100):
+        for i in range(1000):
```
Comment (internaccounts123) [inline simulation/world/world.py:51]: while true

File: simulation/world/world.py [hunk 10]
```diff
@@ -37,46 +48,44 @@ def world_map (self, world_map ):
...
+
+                if self.world_map.is_first_lane_id(car.road_id, car.lane_id):
+                    left_car_list = []
+                else:
+                    left_car_list = self.__grid[car.road_id][car.lane_id - 1]
+                    left_lane_points, left_d_points = self.__world_map.get_lane_points(car.road_id, car.lane_id - 1)
+
+                dec = car.make_decision(self.__grid, lane_points, d_points, right_lane_points,right_d_points,right_car_list, left_lane_points, left_d_points, left_car_list)
+
+                log = Logger.get_logger("FILE")
```
Comment (internaccounts123) [inline simulation/world/world.py:78]: Do not get logger in every iteration.

File: simulation/traffic/vehicle/rule_based/rule_based.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+        next_point = point_to_line_intersection(np.array([self.x, self.y]), np.array([_neigh_1[0], _neigh_2[0]]))
+        car_at_next_point = copy.deepcopy(self)
+
+        car_at_next_point.x = next_point[0]
+        car_at_next_point.y = next_point[1]
+
+        neigh_1, neigh_2 = get_neighbouring_points(side_lane_points, [car_at_next_point.x,car_at_next_point.y])
+        bearing = AngleCalculator.get_bearing(neigh_1[0], neigh_2[0])
+        lower_limit = (car_at_next_point.x - (car_at_next_point.car_length / 2.0) * np.cos(bearing), car_at_next_point.y
```
Comment (internaccounts123) [inline simulation/traffic/vehicle/rule_based/rule_based.py:25]: move this to drive calculations

File: map/map_class.py (hunk 5 above)
Comment (IbtisaamButt) [reply map/map_class.py:60]: Discussed but no changes made.

**Pull request merged**